test/runtests.jl

```diff
@@ -1,125 +1,311 @@
+            gdf1 = groupby(deepcopy(df), [:hod, :obj])
+            gdf2 = groupby(df, [:hod, :obj])
+
+            f1 = x -> Impute.interp(x; context=ctx) |> Impute.locf!() |> Impute.nocb!()
+            f2 = x -> Impute.interp!(x; context=ctx) |> Impute.locf!() |> Impute.nocb!()
+
+            result = vcat(f1.(gdf1)...)
+            @test df != result
+            @test size(result) == (24 * 12 * 10, 3)
+            @test !any(ismissing, Tables.matrix(result))
+
+            # Test that we can also mutate the dataframe directly
+            f2.(gdf2)
+            @test result == sort(df, (:hod, :obj))
+        end
     end
 
     @testset "Matrix" begin
+        ctx = Context(; limit=1.0)
+        expected = Matrix(Impute.dropobs(dataset("boot", "neuro"); context=ctx))
         data = Matrix(dataset("boot", "neuro"))
 
         @testset "Drop" begin
-            result = Iterators.drop(data)
+            result = impute(DropObs(; context=ctx), data)
             @test size(result, 1) == 4
+            @test result == Impute.dropobs(data; context=ctx)
+
+            @test result == expected
+            @test Impute.dropobs(data'; vardim=1, context=ctx) == expected'
         end
 
         @testset "Fill" begin
-            result = impute(data, :fill, 0.0; limit=1.0)
+            result = impute(Fill(; value=0.0, context=ctx), data)
             @test size(result) == size(data)
+            @test result == Impute.fill(data; value=0.0, context=ctx)
+
+            data2 = copy(data)
+            Impute.fill!(data2; value=0.0, context=ctx)
+            @test data2 == result
         end
     end
 
     @testset "Not enough data" begin
-        @test_throws ImputeError impute(a, :drop)
+        ctx = Context(; limit=0.1)
+        @test_throws ImputeError impute(DropObs(; context=ctx), a)
+        @test_throws ImputeError Impute.dropobs(a; context=ctx)
     end
 
     @testset "Chain" begin
-        data = Matrix(dataset("boot", "neuro"))
-        result = chain(
-            data,
-            Impute.Interpolate(),
-            Impute.LOCF(),
-            Impute.NOCB();
-            limit=1.0
-        )
-
-        @test size(result) == size(data)
-        # Confirm that we don't have any more missing values
-        @test !any(ismissing, result)
+        orig = dataset("boot", "neuro")
+        ctx = Context(; limit=1.0)
+
+        @testset "DataFrame" begin
+            result = Impute.interp(orig; context=ctx) |> Impute.locf!() |> Impute.nocb!()
+
+            @test size(result) == size(orig)
+            # Confirm that we don't have any more missing values
+            @test !any(ismissing, Matrix(result))
+
+
+            # We can also use the Chain type with explicit Imputor types
+            result2 = impute(
+                Impute.Chain(
+                    Impute.Interpolate(; context=ctx),
+                    Impute.LOCF(),
+                    Impute.NOCB()
+                ),
+                orig,
+            )
```

Comment:
Maybe something for an issue: `Chain` should be able to have a `Context` that's updated/shared by all of it's imputors?

Reply:
I previously supported passing a context to the `Chain` impute call, but the context limit really only applies to the first imputor anyways because if it passes the first round of imputation then it'll work for all others.